src/serde/object_cache.rs

```diff
@@ -0,0 +1,154 @@
+/// `ObjectCache` provides a way to calculate and cache values for each node
+/// in a clvm object tree. It can be used to calculate the sha256 tree hash
+/// for an object and save the hash for all the child objects for building
+/// usage tables, for example.
+///
+/// It also allows a function that's defined recursively on a clvm tree to
+/// have a non-recursive implementation (as it keeps a stack of uncached
+/// objects locally).
+use crate::allocator::{Allocator, NodePtr, SExp};
+type CachedFunction<T> = fn(&mut ObjectCache<T>, &Allocator, NodePtr) -> Option<T>;
+use super::bytes32::{hash_blobs, Bytes32};
+
+pub struct ObjectCache<'a, T> {
+    cache: Vec<Option<T>>,
+    allocator: &'a Allocator,
+
+    /// The function `f` is expected to calculate its T value recursively based
+    /// on the T values for the left and right child for a pair. For an atom, the
+    /// function f must calculate the T value directly.
+    ///
+    /// If a pair is passed and one of the children does not have its T value cached
+    /// in `ObjectCache` yet, return `None` and f will be called with each child in turn.
+    /// Don't recurse in f; that's part of the point of this function.
+    f: CachedFunction<T>,
+}
+
+/// turn a `NodePtr` into a `usize`. Positive values become even indices
+/// and negative values become odd indices.
+
+fn node_to_index(node: &NodePtr) -> usize {
+    let node = *node;
+    if node < 0 {
+        (-node - node - 1) as usize
+    } else {
+        (node + node) as usize
+    }
+}
+
+impl<'a, T: Clone> ObjectCache<'a, T> {
```

Comment:
this class should have unit tests, but I also have a feeling it could be simplified (made more concrete) to fit the one use we have for it.

Reply:
Added tests. As I've noted previously, the complexity came out of my frustration having written this code several times for different things.